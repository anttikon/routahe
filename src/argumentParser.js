import Moment from 'moment'
import {assignIn} from 'lodash'

exports.parseArguments = (args) => {
  const {isTime, isDate, mapAddresses, parseDateTime} = exports
  const options = args.reduce((opts, arg) => {
    if (arg && isTime(arg)) {
      opts.arriveBy = arg.startsWith('@')
      opts.time = arg.replace('@', '')
    } else if (arg && isDate(arg)) {
      opts.date = arg
    } else if (arg && !opts.address1) {
      opts.address1 = arg
    } else if (arg && opts.address1 && !opts.address2) {
      opts.address2 = arg
    }
    return opts
  }, {})

  if (options.time) {
    options.dateTime = parseDateTime(options)
  }

  return mapAddresses(options)
}

exports.mapAddresses = (options) => {
  const opts = assignIn({}, options)
  const {address1, address2} = opts
  if (address1 && address2) {
    opts.addressFrom = address1
    opts.addressTo = address2
  } else {
    opts.addressTo = address1
  }
  delete opts.address1
  delete opts.address2
  return opts
}


exports.parseDateTime = (opts) => {
  const date = opts.date ? Moment(opts.date, ['DD.MM.YYYY', 'DD.MM.', 'DD.MM', 'DD.M.YYYY', 'DD.M.', 'DD.M', 'D.M', 'D.MM', 'D.MM.', 'D.M.YYYY', 'D.MM.YYYY'], true) : new Moment()
  const time = Moment(opts.time, ['HH:mm', 'H:mm'], true)
  if (!time.isValid()) {
    return null
  }
  time.set({date: date.date(), month: date.month(), year: date.year()})
  return time
}

exports.isTime = (arg) => {
  return arg && !!arg.match(/^@?\d{1,2}:\d{2}$/g)
}

exports.isDate = (arg) => {
  return arg && !!arg.match(/^\d{1,2}\.\d{1,2}\.?(\d{4})?$/g)
}