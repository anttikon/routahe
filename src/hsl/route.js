import fetch from 'node-fetch'

const numItineraries = 5

export const getQueryFromArgs = (args) => {
  const { arriveBy, dateTime, from, to, transports } = args
  const date = dateTime.format('YYYY-MM-DD')
  const time = dateTime.format('HH:mm')
  const transportModes = transports.map(t => ({ mode: t.toUpperCase() }))
  if (transportModes.length > 0) {
    //  All routes require WALK-option
    transportModes.push({ mode: 'WALK' })
  }
  return `
{
  plan(numItineraries: ${numItineraries}, from: {lat: ${from.lat}, lon: ${from.lon}}, to: {lat: ${to.lat}, lon: ${to.lon}}, date: "${date}", time: "${time}", arriveBy: ${arriveBy}, transportModes: ${JSON.stringify(transportModes).replace(/\"/g, '')}) {
    itineraries {
      duration
      legs {
        startTime
        endTime
        duration
        distance
        mode
        from {
          name
        }
        to {
          name
        }
        route {
          shortName
        }
      }
    }
  }
}
`
}

export const getRoutesByQuery = async (query) => {
  const result = await fetch('https://api.digitransit.fi/routing/v1/routers/hsl/index/graphql', {
    method: 'POST',
    body: JSON.stringify({ query }),
    headers: { 'Content-Type': 'application/json' },
  })
  const json = await result.json()
  return json?.data?.plan?.itineraries
}
