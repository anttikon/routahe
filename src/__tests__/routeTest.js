import sinon from 'sinon'

import {validateLocation, action} from '../route'
import {parseArguments} from '../argumentParser'

describe('Route', function() {

  beforeEach(() => sinon.spy(console, 'log'))
  afterEach(() => console.log.restore())

  describe('validateLocation', () => {
    it('should throw error if label is missing', () => {
      expect(() => {
        validateLocation({lon: 12, lat: 32})
      }).toThrow('Location is missing mandatory information: {"lon":12,"lat":32}')
    })
    it('should throw error if lon is missing', () => {
      expect(() => {
        validateLocation({label: 'Steissi', lat: 32})
      }).toThrow('Location is missing mandatory information: {"label":"Steissi","lat":32}')
    })
    it('should throw error if lat is missing', () => {
      expect(() => {
        validateLocation({label: 'Steissi', lon: 12})
      }).toThrow('Location is missing mandatory information: {"label":"Steissi","lon":12}')
    })
    it('should work if route does have label, lon and lat', () => {
      expect(validateLocation({label: 'Steissi', lon: 12, lat: 32})).toBe(true)
    })
  })

  it('action', async() => {
    await action(parseArguments(['Kolmas linja 3, Helsinki', 'kannistontie 2, vantaa', '12.12.2016', '13:32']))
    const args = console.log.getCalls().map(call => call.args)
    expect(args).toMatchSnapshot()
  })

  it('action arrive by', async() => {
    await action(parseArguments(['Suomenlinna, Helsinki', 'Rukatunturintie 2, Helsinki', '1.12.2016', '@13:32']))
    const args = console.log.getCalls().map(call => call.args)
    expect(args).toMatchSnapshot()
  })

})