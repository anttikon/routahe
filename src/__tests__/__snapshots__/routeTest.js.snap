exports[`Route action 1`] = `
Array [
  Array [
    "[1mKolmas linja 3, Helsinki[22m - [1mKannistontie 2, Vantaa[22m | [90mdeparture [39m[36m13:32 12.12.2016[39m",
  ],
  Array [
    "",
  ],
  Array [
    "13:32 - 14:30 58min",
  ],
  Array [
    "[90m  |  [39m13:32 - 13:34 ðŸš¶ [90mWALK[39m [1m   [22m -> Haapaniemi",
  ],
  Array [
    "[34m  |  [39m13:35 - 13:41 ðŸšŒ [34mBUS[39m [1m72  [22m -> Rautatientori",
  ],
  Array [
    "[90m  |  [39m13:41 - 13:46 ðŸš¶ [90mWALK[39m [1m   [22m -> Helsinki",
  ],
  Array [
    "[35m  |  [39m13:51 - 14:15 ðŸš‹ [35mRAIL[39m [1mP  [22m -> KivistÃ¶",
  ],
  Array [
    "[90m  |  [39m14:15 - 14:16 ðŸš¶ [90mWALK[39m [1m   [22m -> KivistÃ¶n asema",
  ],
  Array [
    "[34m  |  [39m14:27 - 14:29 ðŸšŒ [34mBUS[39m [1m443K[22m -> SÃ¤lpÃ¤kuja",
  ],
  Array [
    "[90m  |  [39m14:29 - 14:30 ðŸš¶ [90mWALK[39m [1m   [22m -> Destination",
  ],
  Array [
    "",
  ],
  Array [
    "",
  ],
  Array [
    "13:33 - 14:42 1h 08min",
  ],
  Array [
    "[90m  |  [39m13:33 - 13:35 ðŸš¶ [90mWALK[39m [1m [22m -> Haapaniemi",
  ],
  Array [
    "[34m  |  [39m13:36 - 13:42 ðŸšŒ [34mBUS[39m [1m71[22m -> Rautatientori",
  ],
  Array [
    "[90m  |  [39m13:42 - 13:44 ðŸš¶ [90mWALK[39m [1m [22m -> Helsinki",
  ],
  Array [
    "[35m  |  [39m13:51 - 14:25 ðŸš‹ [35mRAIL[39m [1mI[22m -> KivistÃ¶",
  ],
  Array [
    "[90m  |  [39m14:25 - 14:42 ðŸš¶ [90mWALK[39m [1m [22m -> Destination",
  ],
  Array [
    "",
  ],
  Array [
    "",
  ],
  Array [
    "13:40 - 14:42 1h 01min",
  ],
  Array [
    "[90m  |  [39m13:40 - 13:43 ðŸš¶ [90mWALK[39m [1m [22m -> Kallion virastotalo",
  ],
  Array [
    "[34m  |  [39m13:44 - 13:58 ðŸšŒ [34mBUS[39m [1m23[22m -> Pasilan asema",
  ],
  Array [
    "[90m  |  [39m13:58 - 13:59 ðŸš¶ [90mWALK[39m [1m [22m -> Pasila",
  ],
  Array [
    "[35m  |  [39m14:05 - 14:25 ðŸš‹ [35mRAIL[39m [1mP[22m -> KivistÃ¶",
  ],
  Array [
    "[90m  |  [39m14:25 - 14:42 ðŸš¶ [90mWALK[39m [1m [22m -> Destination",
  ],
  Array [
    "",
  ],
]
`;

exports[`Route action arrive by 1`] = `
Array [
  Array [
    "[1mSuomenlinna, Helsinki[22m - [1mRukatunturintie 2, Helsinki[22m | [90marrive [39m[36m13:32 01.12.2016[39m",
  ],
  Array [
    "",
  ],
  Array [
    "12:15 - 13:30 1h 14min",
  ],
  Array [
    "[90m  |  [39m12:15 - 12:19 ðŸš¶ [90mWALK[39m [1m   [22m -> Suomenlinna, pÃ¤Ã¤lait",
  ],
  Array [
    "[36m  |  [39m12:20 - 12:35 â›´ [36mFERRY[39m [1m19[22m -> Kauppatori",
  ],
  Array [
    "[90m  |  [39m12:35 - 12:39 ðŸš¶ [90mWALK[39m [1m   [22m -> Kauppatori",
  ],
  Array [
    "[32m  |  [39m12:50 - 12:54 ðŸšŠ [32mTRAM[39m [1m2  [22m -> Mikonkatu",
  ],
  Array [
    "[90m  |  [39m12:54 - 12:59 ðŸš¶ [90mWALK[39m [1m   [22m -> Helsingin yliopisto",
  ],
  Array [
    "[33m  |  [39m13:02 - 13:20 ðŸš‡ [33mSUBWAY[39m [1m [22m -> MellunmÃ¤ki",
  ],
  Array [
    "[90m  |  [39m13:20 - 13:22 ðŸš¶ [90mWALK[39m [1m   [22m -> MellunmÃ¤ki(M)",
  ],
  Array [
    "[34m  |  [39m13:25 - 13:29 ðŸšŒ [34mBUS[39m [1m97  [22m -> KaunispÃ¤Ã¤ntie",
  ],
  Array [
    "[90m  |  [39m13:29 - 13:30 ðŸš¶ [90mWALK[39m [1m   [22m -> Destination",
  ],
  Array [
    "",
  ],
  Array [
    "",
  ],
  Array [
    "11:35 - 12:49 1h 13min",
  ],
  Array [
    "[90m  |  [39m11:35 - 11:39 ðŸš¶ [90mWALK[39m [1m   [22m -> Suomenlinna, pÃ¤Ã¤lait",
  ],
  Array [
    "[36m  |  [39m11:40 - 11:55 â›´ [36mFERRY[39m [1m19[22m -> Kauppatori",
  ],
  Array [
    "[90m  |  [39m11:55 - 11:59 ðŸš¶ [90mWALK[39m [1m   [22m -> Ritarihuone",
  ],
  Array [
    "[32m  |  [39m12:04 - 12:08 ðŸšŠ [32mTRAM[39m [1m4  [22m -> Ylioppilastalo",
  ],
  Array [
    "[90m  |  [39m12:08 - 12:12 ðŸš¶ [90mWALK[39m [1m   [22m -> Rautatientori",
  ],
  Array [
    "[33m  |  [39m12:15 - 12:35 ðŸš‡ [33mSUBWAY[39m [1m [22m -> MellunmÃ¤ki",
  ],
  Array [
    "[90m  |  [39m12:35 - 12:38 ðŸš¶ [90mWALK[39m [1m   [22m -> MellunmÃ¤ki (M)",
  ],
  Array [
    "[34m  |  [39m12:43 - 12:47 ðŸšŒ [34mBUS[39m [1m560 [22m -> Fallpakka",
  ],
  Array [
    "[90m  |  [39m12:47 - 12:49 ðŸš¶ [90mWALK[39m [1m   [22m -> Destination",
  ],
  Array [
    "",
  ],
  Array [
    "",
  ],
  Array [
    "10:35 - 11:49 1h 13min",
  ],
  Array [
    "[90m  |  [39m10:35 - 10:39 ðŸš¶ [90mWALK[39m [1m   [22m -> Suomenlinna, pÃ¤Ã¤lait",
  ],
  Array [
    "[36m  |  [39m10:40 - 10:55 â›´ [36mFERRY[39m [1m19[22m -> Kauppatori",
  ],
  Array [
    "[90m  |  [39m10:55 - 10:59 ðŸš¶ [90mWALK[39m [1m   [22m -> Ritarihuone",
  ],
  Array [
    "[32m  |  [39m11:04 - 11:08 ðŸšŠ [32mTRAM[39m [1m4  [22m -> Ylioppilastalo",
  ],
  Array [
    "[90m  |  [39m11:08 - 11:12 ðŸš¶ [90mWALK[39m [1m   [22m -> Rautatientori",
  ],
  Array [
    "[33m  |  [39m11:15 - 11:35 ðŸš‡ [33mSUBWAY[39m [1m [22m -> MellunmÃ¤ki",
  ],
  Array [
    "[90m  |  [39m11:35 - 11:38 ðŸš¶ [90mWALK[39m [1m   [22m -> MellunmÃ¤ki (M)",
  ],
  Array [
    "[34m  |  [39m11:43 - 11:47 ðŸšŒ [34mBUS[39m [1m560 [22m -> Fallpakka",
  ],
  Array [
    "[90m  |  [39m11:47 - 11:49 ðŸš¶ [90mWALK[39m [1m   [22m -> Destination",
  ],
  Array [
    "",
  ],
]
`;
